/*
 * This file is part of the NUbots Codebase.
 *
 * The NUbots Codebase is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The NUbots Codebase is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the NUbots Codebase.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Copyright 2013 NUBots <nubots@nubots.net>
 */

syntax = "proto3";

package messages.vision;

import "messages/MessageOptions.proto";
import "messages/Geometry.proto";
import "messages/Vector.proto";
import "messages/input/Image.proto";
import "messages/input/Sensors.proto";

message ClassifiedImage {
    enum ObjectClass {
        UNKNOWN = 0;
        FIELD = 1;
        BALL = 2;
        GOAL = 3;
        LINE = 4;
        CYAN_TEAM = 5;
        MAGENTA_TEAM = 6;
    }

    message Segment {
        ObjectClass colour = 1;
        uint32 length = 2;
        uint32 subsaple = 3;

        ivec2 start = 4;
        ivec2 end = 5;
        ivec2 midpoint = 6;

        Segment previous = 7 [(pointer)=RAW];
        Segment next = 8 [(pointer)=RAW];
    }



    // The sensors that were used for this classified image
    input.Sensors sensors = 1 [(pointer)=SHARED];
    // The image used for this classified image
    input.Image image = 2 [(pointer)=SHARED];

    ivec2 dimensions = 3;

    Line horizon = 4;
    Polyline visual_horizon = 5;

    map<uint32, Segment> horizontal_segments = 6;
    map<uint32, Segment> vertical_segments = 7;


    // High number because this is a hack...
    // Seed points for the lightning ball model
    message BallSeedPoints {
        repeated ivec2 point = 1;
    }
    repeated BallSeedPoints ball_seed_points = 1000;
    repeated ivec2 ball_points = 1001;
}
